### PR TITLE
Initial Spans/Pipelines port

### PR DESCRIPTION
**Pull request by ReubenBond**

### CODE REVIEW

File: src/Hagar/Buffers/Writer.cs [hunk 1]
```diff
@@ -1,163 +1,166 @@
...
-            this.Trace("--Wrote decimal {0}", d);
-            this.ab.Append(Decimal.GetBits(d));
+            var ints = Decimal.GetBits(d);
+            foreach (var part in ints) this.Write(part);
         }
 
-        // Text
-
         /// <summary> Write a <c>string</c> value to the stream. </summary>
         public void Write(string s)
```
Comment (davidfowl) [inline src/Hagar/Buffers/Writer.cs]: This logic isn't very efficient. You can use the UT8 decoder APIs to decode directly into the `IBufferWriter<byte>`.

File: src/Hagar/Buffers/Writer.cs [hunk 2]
```diff
@@ -1,163 +1,166 @@
...
         public void Write(byte[] b)
         {
-            this.Trace("--Wrote byte array of length {0}", b.Length);
-            this.ab.Append(b);
+            this.Write(new ReadOnlySpan<byte>(b));
         }
 
-        /// <summary> Write a list of byte array segments to the stream. </summary>
-        public void Write(List<ArraySegment<byte>> bytes)
+        public void Write(ReadOnlySpan<byte> input)
```
Comment (davidfowl) [inline src/Hagar/Buffers/Writer.cs]: This extension method already exists in the System.Buffers namespace https://docs.microsoft.com/en-us/dotnet/api/system.buffers.buffersextensions.write?view=netcore-2.1#System_Buffers_BuffersExtensions_Write__1_System_Buffers_IBufferWriter___0__System_ReadOnlySpan___0__. Thought this does interface dispatch.

File: src/Hagar/Utilities/VarIntWriterExtensions.cs [hunk 3]
```diff
@@ -12,50 +12,50 @@ public static class VarIntWriterExtensions
 
         public static void WriteVarInt(this Writer writer, byte value)
         {
-            var scratch = new byte[2];
+            Span<byte> scratch = stackalloc byte[2];
```
Comment (davidfowl) [inline src/Hagar/Utilities/VarIntWriterExtensions.cs]: These methods are copying. Why not write directly into the writer?

File: src/Hagar/Buffers/Reader.cs [hunk 4]
```diff
@@ -1,432 +1,317 @@
...
+using System.Runtime.InteropServices;
 using System.Text;
+using Hagar.Utilities;
 
 namespace Hagar.Buffers
 {
-    /// <summary>
-    /// Reader for Orleans binary token streams
-    /// </summary>
     public class Reader
```
Comment (davidfowl) [inline src/Hagar/Buffers/Reader.cs]: The Reader is currently copying a bit too much. You want to keep track of the current memory and drain it until it's empty. Then move to the next memory when you run out. That's what the google protobuf prototype does:

https://github.com/mkosieradzki/protobuf/blob/b6a270101811a1ca6c3a4a476df7028fde93a228/csharp/src/Google.Protobuf/CodedInputStream.cs#L89-L92

Fast path:

https://github.com/mkosieradzki/protobuf/blob/b6a270101811a1ca6c3a4a476df7028fde93a228/csharp/src/Google.Protobuf/CodedInputStream.cs#L1350-L1362

Fast path has no copies, uses the current span.

Slow path:

https://github.com/mkosieradzki/protobuf/blob/b6a270101811a1ca6c3a4a476df7028fde93a228/csharp/src/Google.Protobuf/CodedInputStream.cs#L1366-L1373

Reads a byte at a time.

Comment (ReubenBond) [reply src/Hagar/Buffers/Reader.cs]: @davidfowl how about now? I haven't updated Writer yet, but I'll give it the same treatment if this is on the right track.

File: src/Hagar/Buffers/Reader.cs [hunk 5]
```diff
@@ -1,432 +1,231 @@
...
+        public Reader ForkFrom(long position)
         {
-            this.buffers = buffs;
-            this.totalProcessedBytes = 0;
-            this.currentSegmentIndex = 0;
-            this.InitializeCurrentSegment(0);
-            this.totalLength = buffs.Sum(b => b.Count);
-            this.buffersCount = buffs.Count;
+            var result = new Reader(this.input);
+            result.Skip(position);
```
Comment (davidfowl) [inline src/Hagar/Buffers/Reader.cs]: I would just slice the buffer before passing it to the reader.

**Comment by ReubenBond** [comment]
This PR is just dipping my toes into the water.

I'm redoing the `Writer` now. After that, I'm going to adjust the codec interfaces and update codegen. At that point it will be fully spannified.

There will still a significant amount of performance left hanging on the vine. I'll address that afterwards.

Thank you, @davidfowl

**Comment by davidfowl** [comment]
Nice! It’s looking good. One thing to consider is since this is all synchronous if you want to make the writer and reader a ref struct (and therefore stack only) and store the current span in a field instead of memory. 

I’d do that after getting performance tests setup

**Comment by ReubenBond** [comment]
@davidfowl I took the ref struct approach in the latest commit. How do Reader & Writer look now?

My current benchmark only compares to the existing Orleans serializer. It does not compare favorably in most cases yet. It is still unoptimized and also has significantly more work to do since it's a version-tolerant serializer whereas the inbuilt Orleans serializer is not.

|              Method |        Mean |      Error |     StdDev |  Gen 0 | Allocated |
|-------------------- |------------:|-----------:|-----------:|-------:|----------:|
|     HagarSerializer | 12,376.3 ns | 238.130 ns | 274.231 ns | 0.9003 |    3824 B |
|   OrleansSerializer |  8,158.8 ns | 160.102 ns | 224.441 ns | 1.4496 |    6128 B |
|      HagarSerialize |  4,884.1 ns |  97.064 ns |  90.793 ns | 0.7172 |    3016 B |
|    OrleansSerialize |  4,745.5 ns |  95.323 ns | 161.866 ns | 1.2054 |    5088 B |
|    HagarDeserialize |  6,540.7 ns |  69.225 ns |  61.366 ns | 0.1907 |     808 B |
|   HagarReadEachByte |    524.1 ns |   9.617 ns |   8.995 ns |      - |       0 B |
| OrleansReadEachByte |    564.5 ns |  11.211 ns |  13.346 ns | 0.0315 |     136 B |
|  OrleansDeserialize |  2,639.3 ns |  49.166 ns |  43.584 ns | 0.2251 |     952 B |

**Comment by davidfowl** [comment]
Do you know where the allocations in HagarDeserialize come from?

**Comment by ReubenBond** [comment]
These are the allocs (ignore the precise numbers: that's from running a tight serialize/deserialize loop):
image

The class being serialized is "complex" and forces the serializer to hit expensive paths (type name serialization), and that's what's being seen there: type name serialization. My plan is to create a more efficient representation of the .NET Type system, but the library also allows "well known types" and "referenced types" for more efficient serialization. I'm just not using them in this test.

If we change the type to a simple struct (an int, a bool, and an `object` field which is always null) then we see better results:

|                 Method |     Mean |     Error |    StdDev |  Gen 0 | Allocated |
|----------------------- |---------:|----------:|----------:|-------:|----------:|
|   HagarStructRoundTrip | 1.169 us | 0.0225 us | 0.0300 us | 0.0114 |      56 B |
| OrleansStructRoundTrip | 2.407 us | 0.0469 us | 0.0743 us | 1.2207 |    5136 B |

image
Clearly still room for improvement: no need to call `Activator.CreateInstance<T>` for a value type.

**Comment by ReubenBond** [comment]
I assume it implies I'm leaking memory, maybe because I'm using the pipe incorrectly at the top level of the benchmark code?

```C#
private readonly Pipe sharedPipe = new Pipe(new PipeOptions(useSynchronizationContext: false));
[Benchmark]
public SimpleStruct HagarStructRoundTrip()
{
    var writer = new Writer(sharedPipe.Writer);
    session.FullReset();
    this.structSerializer.Serialize(this.structValue, session, ref writer);

    session.FullReset();
    sharedPipe.Writer.FlushAsync().GetAwaiter().GetResult();
    var readResult = sharedPipe.Reader.ReadAsync().GetAwaiter().GetResult();
    var reader = new Reader(readResult.Buffer);
    var result = this.structSerializer.Deserialize(session, ref reader);
    sharedPipe.Reader.AdvanceTo(readResult.Buffer.End);
    return result;
}
```

**Comment by ReubenBond** [comment]
Some optimizations for value types:

|                 Method |       Mean |    Error |   StdDev |  Gen 0 | Allocated |
|----------------------- |-----------:|---------:|---------:|-------:|----------:|
|   HagarStructRoundTrip |   946.7 ns | 11.70 ns | 10.37 ns | 0.0124 |      56 B |
| OrleansStructRoundTrip | 2,247.9 ns | 44.50 ns | 75.56 ns | 1.2283 |    5168 B |

Still many low hanging fruit

**Comment by davidfowl** [comment]
> I assume it implies I'm leaking memory, maybe because I'm using the pipe incorrectly at the top level of the benchmark code?

Right, if you never consume the data then it's a leak. Also calling Complete on both pipes should return the memory. The array pool does allocate an object per rent which wraps the underlying array (which is a bit unfortunate).

You shouldn't need a dependency on pipelines at all in the main product. You're only using it in tests. Unless you plan to add an async deserialize.

**Comment by davidfowl** [comment]
Is this xunit? Call Complete on both pipes in dispose of the class.

**Comment by ReubenBond** [comment]
Is the data not being consumed in that benchmark code above? I thought that's what AdvanceTo was doing.

I doubt I'll be adding async deserialization any time soon, I don't thinks it's worthwhile. Better to just length-prefix payloads.

**Comment by davidfowl** [comment]
> Is the data not being consumed in that benchmark code above? I thought that's what AdvanceTo was doing.

Yes it is, so this is likely fine. I would still call complete.

> I doubt I'll be adding async deserialization any time soon, I don't thinks it's worthwhile. Better to just length-prefix payloads.

Remove the pipelines dependency from the product and keep it in the app and benchmark.

**Pull request merged**